### ISSUE TITLE
FR: Creation or reconfiguration of Gestures and Profiles  based on current reading states.

### ISSUE THREAD

**Issue by solarkite** [koreader/koreader#8590]
Creating or reconfiguring Gestures and Profiles can be cumbersome, and, if complex, difficult to replicate.
Also there seems to be a desire for users to match Gestures with Profiles. 
Ideally, It would be desirable to match a gesture with a profile. 
The concern that seems to be raised is that if a plug-in fails or is disabled, then having it married to another plug-in could cause the other plug-in to have problems. An argument for plug-in independence is offered.
So, If a pointer from gestures to profiles, and/or vice-verse, cannot be enabled, 
then perhaps a creation or reconfiguration of Gestures or Profiles built from current reading states could allow, at the least, a quick replication of many reading settings to closely mirror such in another. 
For Profiles, the "Profile" menu below the "New" selection add "New (with current document settings)"
And within each profile, below "Delete Profile" , add the selection  "Set values at current document conditions"  (Or alternatively, "Use current document settings")
Likewise, within the "Gesture Manager" within the selections for each gesture at the same level as "Reflowable documents" and "Fixed layout documents" add to the top selection  "Set values at current document conditions" (Or alternatively, "Use current document settings")

**Comment by solarkite** [comment]
![
![New with current document settings](https://user-images.githubusercontent.com/95706893/147415771-2e495703-ecae-46a9-bb35-823ae10018f5.png)
set values - Gestures](https://user-images.githubusercontent.com/95706893/147415727-4f9f53c0-fd7e-458d-af8b-c4e62f66f79d.png)
![Profile - set values](https://user-images.githubusercontent.com/95706893/147415728-4c00628e-21f0-45a3-9b2b-8211093846ac.png)

**Comment by solarkite** [comment]
@hius07 
Given the extensive work you’ve been doing in profiles and gestures lately, I thought I’d bring attention to this suggestion to see what you think of implementation. I could see benefits to quickly recording and replicating reading conditions without fiddling around with every setting again.

**Comment by NiLuJe** [comment]
That would require massively reworking how settings themselves work.

The default value of a setting is not actually stored anywhere, as such, we can't retrieve it. So we can't really know what's been modified.
Also, there's no real hierarchy, it's just (essentially) a bunch of key, value pairs; so you can't walk the setting table and say: "oh, this belongs *here*".

So, there's already a very basic problem here, and that's not even getting into the Dispatcher integration *at all*.

It works in practice because the code that actually *uses* the setting knows what to do with it, but that's it.
(And, even there, handling of defaults is... sometimes clunky).

(See also "why settings cannot be guaranteed to be portable", which was discussed recently).

**Comment by hius07** [comment]
Estimated list of document settings for the purpose of this issue:

![01](https://user-images.githubusercontent.com/62179190/191966119-d1cbe9b4-676b-40aa-b310-ee584d3945fe.png)

![02](https://user-images.githubusercontent.com/62179190/191966145-29a20679-d4e7-4938-a7a1-884c9bcf33c4.png)

So, the module would get the current document settings (if a document setting is not set - take or invent the default value) and generate the profile with the above list of actions.

**Comment by solarkite** [comment]
> 



> the module would get the current document settings (if a document setting is not set - take or invent the default value) and generate the profile with the above list of actions.

An interesting take on the idea. I originally thought about saving it as a standard profile. But you seem to have taken it one step further to suggest a quick menu option to be executed. That could work. You wouldn’t have to view the whole action sequence to trigger it. I wonder if using a gesture to save the current document settings under a new quick menu would be easy?

**Comment by hius07** [comment]
I showed the quick menus just to list the settings, maybe some more settings should be included?
The "Current document settings" profile will be a standard profile that can be executed from the profiles list or with a gesture or shown as a QuickMenu (the profiles system is currently under development in https://github.com/koreader/koreader/pull/9531).

**Comment by hius07** [comment]
While we are waiting for https://github.com/koreader/koreader/pull/9531 I've made the following list:

```Lua
function Profiles:getProfileFromCurrentDocument()
    local profile
    if self.ui.rolling then
        profile = {
            ["set_font"] = self.ui.font.font_face,
            ["font_size"] = self.ui.font.font_size,
            ["font_base_weight"] = self.ui.font.font_base_weight,
            ["font_hinting"] = self.ui.font.font_hinting,
            ["font_kerning"] = self.ui.font.font_kerning,
            ["font_gamma"] = self.ui.font.gamma_index,
            ["line_spacing"] = self.ui.font.line_space_percent,
            ["h_page_margins"] = { self.ui.typeset.unscaled_margins[1],
                                   self.ui.typeset.unscaled_margins[3] },
            ["t_page_margin"] = self.ui.typeset.unscaled_margins[2],
            ["b_page_margin"] = self.ui.typeset.unscaled_margins[4],
            ["block_rendering_mode"] = self.ui.typeset.block_rendering_mode,
            ["view_mode"] = self.view.view_mode,
            ["visible_pages"] = self.ui.rolling.visible_pages,
        }
    else
        local trim_page_to_mode = {
            [3] = "none",
            [2] = "semi-auto",
            [1] = "auto",
            [0] = "manual",
        }
        local zoom_genus_to_mode = {
            [4] = "page",
            [3] = "content",
            [2] = "columns",
            [1] = "rows",
            [0] = "manual",
        }
        local zoom_type_to_mode = {
            [2] = "full",
            [1] = "width",
            [0] = "height",
        }
        profile = {
            ["kopt_contrast"] = self.document.configurable.contrast,
            ["kopt_font_size"] = self.document.configurable.font_size,
            ["kopt_line_spacing"] = self.document.configurable.line_spacing,
            ["kopt_max_columns"] = self.document.configurable.max_columns,
            ["kopt_page_margin"] = self.document.configurable.page_margin,
            ["kopt_quality"] = self.document.configurable.quality,
            ["kopt_text_wrap"] = self.document.configurable.text_wrap,
            ["kopt_trim_page"] = trim_page_to_mode[self.document.configurable.trim_page],
            ["kopt_zoom_direction"] = self.document.configurable.zoom_direction,
            ["kopt_zoom_factor"] = self.document.configurable.zoom_factor,
            ["kopt_zoom_mode_genus"] = zoom_genus_to_mode[self.document.configurable.zoom_mode_genus],
            ["kopt_zoom_mode_type"] = zoom_type_to_mode[self.document.configurable.zoom_mode_type],
            ["kopt_zoom_overlap_h"] = self.document.configurable.zoom_overlap_h,
            ["kopt_zoom_overlap_v"] = self.document.configurable.zoom_overlap_v,
            ["kopt_page_scroll"] = self.view.page_scroll,
        }
    end
    return profile
end

```
It produces the following profiles with the settings of the currently opened document:

![03](https://user-images.githubusercontent.com/62179190/192345071-7d2bb798-24ef-40b7-a3ba-68d0ece2910f.png)

![04](https://user-images.githubusercontent.com/62179190/192345089-e25f7a32-d136-4e13-8f9e-c1510adf78ea.png)

Please comment if any settings shoudl be added to or deleted from the list.

**Comment by solarkite** [comment]
Reviewing document settings has brought to my attention for the first time (so maybe it doesn’t matter) that reflowable documents don’t seem to have a setting for alignment. ¯\_(ツ)_/¯

**Comment by hius07** [comment]
Look at the menu: typeset -> Style tweaks -> Text -> Text alignment.

**Comment by hius07** [comment]
As you can see, currently style tweaks can be applied via the menu only, not with a gesture. So they cannot be added to a profile (Profiles use the same module Dispatcher as the Gesture Manager does).
We can ask @poire-z's opinion on adding Style Tweaks to the Dispatcher.

**Comment by solarkite** [comment]
I see. It does seem like a viable improvement in order to replicate the fullness of a document’s reading state.

**Comment by solarkite** [comment]
> There's a _lot_ of style tweaks, though, and some can even be custom, so I'm not convinced of the practicality of this... Unless we filter the list and only keep some of 'em, but then you're back to choosing what makes the cut... ;o).

I was thinking similarly to some of the style tweaks. Alignment and text direction seem practical in terms of adding them to profiles, and subsequently, the list of "Current document settings" to be saved

**Comment by poire-z** [comment]
Mixed opinions and feelings :)

1) I am/was quite not interested by this feature request (and its buddy: having QuickMenus): it feels like another mindset that I can't share or encourage: _I don't want to understand your menus and settings: allow me to build my own menus with the things (I think) I care about - and to save all the settings in a big bag of crap_. I believe we did things right, and the learning curve to get into KOReader is valuable and needed to make the most of it - and you can take your time to discover it. These 2 feature requests are for the lazy :) and don't encourage discovery - and I don't care if these people abanndon KOReader. I also don't think we need to have KOReader support 2 different mindsets (the learning one, and the lazy one). So, not much to say until now about this FR.
I still think it's better to think about the settings you have tweaked, and manually add them "by topic" in one or multiple profiles - instead of having a big bag of settings you don't even know what differentiates it from that other big bag of settings, and get to find which is this tiny hidden setting that makes this profile a bit more confortable than the other one... Manually thinking about all this get you to know and understand what makes what.
So, not much opinion about the above screenshots - but just happy the code gets mostly added into a plugin - and does not leak much into core.

2) it turns out QuickMenu get to be a nice addition - I'm still not using it, but I understand how it can be useful even for KOReader lovers.

3) I can see a (good, not lazy :) use case for this, though: our "settings set as defaults" don't apply on past books, only on future newly opened books. I can see this gathering of all the current settings in a bag with a name, and this can be think of as different sets of defaults - that can be applied on previously opened books. It still feels dirty and obscure - but so is our single set of defaults settings.
But still, there's many things that makes the "fullness of a document reading state": the ones you added in the above screenshots, + styletweaks, would still be incomplete (missing Style epub/html5.css, typography settings, ToC settings...) so it will always be an incomplete bag.

4) I could think of settings depending on the user mood (font size, family, nightmode, even margins) - may be all the ones in the above screenshots. But I also think some of these are orthogonal to each others (ex. nightmode+contrats vs. reflow mode or 2-columns - you'd need 4 profiles to get the 2x2 combinations of these). So, these many big bags of crap are not the solution to user mood swings.

5) some settings are not really dependant on the mood of the user, but are dependant on the book, and you set/select them at opening time when you prepare the book for your reading. Styletweaks are part of that 2nd set - and unless you read many books published by the same publisher at the same period, one always have to see how the new book looks to decide which tweaks are needed. Having these in a profile associated to a gesture for quick apply a bunch of tweaks is a bit evil/lazy :) You most often will see something odd, and need to go back into the menu and try the other tweaks around. So, I think it's a bad idea to add current tweaks with all the other settings into that bag with a name.
(Text alignment left/justified can be considered "user mood" - and it's a bottom menu button for PDF - but needs to be a styletweak for CRE documents - for technical tidyness. If you want to add styletweaks just for that: don't. text alignment is not really "user mood", rather "user taste" and won't change with the mood - it's a hard preference from what I can read on mobileread forums :)
But it may be useful to have a "set of tweaks", when you know at first opening of a book, that you will need to apply these 4/5 common ones you always need to apply (that's what I often end up doing :/ I don't want to set these 4 tweaks as default, to at least see once the publisher genuine style - and then, I kill his things to make them my taste).
Also, contrary to our other defaults, styletweaks set as default will apply to previously opened books.

6) but, with your recent QuickMenus, I think it could be nice to be able to gather these most commonly used tweaks (different for each user) into a QuickMenu, to quickly apply these 4/5 common ones (I'm so used to the top menu that I may end up not using it, but it would be a nice thing to have :)

7) it would be 96 items (96 shipped styletweaks, + user ones) to register into Dispatcher at each book opening time (or only at the first one?) - and I have no idea how expensive this makes it for Dispatcher. Also, you would need to add the same hierarchical submenus into Dispatcher action menu. Feels not fun, messy, hard to make friendly - but I dunno.
This could also be a solution to the "common sets of styletweaks" (to apply all at once) that would otherwise be hard to implement via our top menu.

8) once all this is implemented for these good reasons, I couldn't stop @solarkite to use all these facilities for his evil reasons :) but we would then not even have to know :)

TLDR: no strong opinion - not strictly necessary, but it could solve some legitimate use cases (and if it solves evil use cases, well, we don't need to know :)
(I also get the feeling @yparitcher may not be super happy getting his Dispatcher invaded with all that new stuff :)

**Comment by hius07** [comment]
We can add a button "Register action" to the TweakInfoWidget to allow a user to register his favourite style tweaks only. To avoid registering of all 96 tweaks.

**Comment by poire-z** [comment]
> We can add a button "Register action" to the TweakInfoWidget to allow a user to register his favourite style tweaks only. To avoid registering of all 96 tweaks.

Sounds like a nice and clean solution - a bit hidden but available to the people who needs it.

(But this may not solve your original problem of saving the "full reading state" including tweaks - if you really want to go on with it :) : would you save only these user-manually-registered tweaks ?)

**Comment by poire-z** [comment]
> (Text alignment left/justified can be considered "user mood" - and it's a bottom menu button for PDF - but needs to be a styletweak for CRE documents - for technical tidyness.

I feel the need to say more about that:

In PDF, when enabling reflow, we get plain lines of text , word boxes stacked together - and we lose all PDF publisher layout, which is the purpose of reflow. So, having buttons to decide how to stack/align them makes sense.

With EPUBs, publishers set their layout with CSS - and we never kill it all (but a user can with the radical `Embedded styles off`.
Publishers decide which kind of text container (titles, paragraphs, table cells..) gets which alignment.
Users may not like their choices, and can modify some of them.
We provide 2 style tweaks for left alignment:
`Left align most text`  - which targets paragraphs only, keeping everything else as-is.
`Left align all text`  - which targets every text container.

If we would want to have a one-stop bottom button to toggle left/justify text, we'd need to decide which of the above to use - and there's no obvious choice.
`Left align most text` may have no effect (if the publisher uses `<DIV>` for the text containers.
`Left align all text` may have a too strong effect (killing centered titles, table columns headers, caption under images...)
I don't know which solution the other ereader softwares use - and I don't think any of these 2 has prevalence.
I do think the user needs to understand how that works, how styles are used in books (and we try to not make it too technical with explanations about a style tweak effect in plain english, but we show the CSS so he can learn), and for him to make the decision.
And if at first, a user may feel we are missing a text alignment left|justify button, I do think it's ok to not provide this - and let him apprehend how this works and what tools he has and how to use them.

**Comment by poire-z** [comment]
Btw, @hius07 @yparitcher : I haven't checked, but I believe that, with our current code, if you apply a profile that changes N cre settings (say font family, font size, word spacing, margin), you will get N re-renderings (so, on a big book, possibly N progress bars at top left) - while we would expect only a re-rendering once all the settings have been applied.
This is due to each eventhandler doing `self.ui:handleEvent(Event:new("UpdatePos"))` as all events were mostly always sent each individually (when we add a feature and a toggle/menu item) - so each making sure to check and ensure a rerendering happens.
I don't even want to think about how to solve this :) if it should be a more generic/global change to how we ensure these rerenderings. Or if it could be enough for Dispatcher to flag something in ReaderRolling before sending all events to prevent these re-rendering in onUpdatepos, and Dispatcher would itself unset the flag and do a final onUpdatePos.

Unless this is already thought of in Dispatcher/Profiles ?

**Comment by hius07** [comment]
Yes, I've got two upper progress bars one by one when made a gesture with two actions (increase font size, set line spacing).

**Comment by yparitcher** [comment]
@poire-z 
I think i mentioned it when designing dispatcher, but it was not a serious usecase so no one asked to optimize it, so i left it as is.


> 1. I am/was quite not interested by this feature request (and its buddy: having QuickMenus): it feels like another mindset that I can't share or encourage: _I don't want to understand your menus and settings: allow me to build my own menus with the things (I think) I care about - and to save all the settings in a big bag of crap_. I believe we did things right, and the learning curve to get into KOReader is valuable and needed to make the most of it - and you can take your time to discover it. These 2 feature requests are for the lazy :) and don't encourage discovery - and I don't care if these people abanndon KOReader. I also don't think we need to have KOReader support 2 different mindsets (the learning one, and the lazy one). So, not much to say until now about this FR.
>        I still think it's better to think about the settings you have tweaked, and manually add them "by topic" in one or multiple profiles - instead of having a big bag of settings you don't even know what differentiates it from that other big bag of settings, and get to find which is this tiny hidden setting that makes this profile a bit more confortable than the other one... Manually thinking about all this get you to know and understand what makes what.
>        So, not much opinion about the above screenshots - but just happy the code gets mostly added into a plugin - and does not leak much into core.

:+1:

**Comment by solarkite** [comment]
@hius07 I was curious about the implementation of the project. Were the results harder to produce that initially projected?

**Comment by hius07** [comment]
There is a new menu item "Save current document settings as default".

**Comment by zmixnyc** [comment]
> There is a new menu item "Save current document settings as default".

It would be nice to have "Save current document settings as... ", and allow the user to name these setups, and recall them.. I had opened an FR[ here](https://github.com/koreader/koreader/issues/10015) about it recently.

**Comment by solarkite** [comment]
> There is a new menu item "Save current document settings as default".

I appreciate the work you put Into that. It’s not a function that I, myself, found a use for. I thought you had the solution to the originally stated problem in this FR: “Creating or reconfiguring Gestures and Profiles can be cumbersome, and, if complex, difficult to replicate.” I am baffled by another commenter referring to my use of such a feature as evil. I think I laid out the problem and a reasonable solution that was able to be implemented. I believe @hius07 thought it was reasonable too.

**Comment by poire-z** [comment]
> It would be nice to have "Save current document settings as... ", and allow the user to name these setups

That's what I referred as "a big bag of crap with a name".

> I am baffled by another commenter referring to my use of such a feature as evil. I think I laid out the problem and a reasonable solution 

That was me, at https://github.com/koreader/koreader/issues/8590#issuecomment-1259955165.
'evil', because everytime you would be using that unreasonable solution, a kitten would be killed!
And unreasonable because of what I explained in item 1).
You think it is, but it is not: we have 90+ settings, should we really save them all and pick the current book language typography/hyphenation settings, and the style tweaks you needed for this book only ? If we do that, applying your bag of settings will do more than what you or any user naively (and rightly) expect.
So, we would need to decide - all of us with various taste/appreciations - what settings can go in and which should be left out - and think about it everytime we add a new setting - and I'd rather not have that.

Quoting @zmixnyc  from #10015 : 

> It's important to view this from the user's perspective, in my case: "I have set up all my font and layout preferences and I really like the way this book looks now, I'd like to be able to apply this to future books

Which is what @solarkite also wants. But then, "font and layout preferences" may be different in the mind of different users, and you're back to my above point.

>  it would be useful if it had an option to "save all current user preference settings" when creating a new profile, and then the user could go through all of the various menus and disable those that they don't need, rather than making them create a profile and then manually change each parameter. It seems backwards to me as a user.

I don't know what you mean with "could go through all of the various menus and disable those that they don't need", if you mean the regular top and bottom menu after that 90+settings profile is applied - in which case it would be even more backwards: you apply that big bag of 90+ crap settings, you get your 12 "font and layout settings" you wanted, but you need to go "fix" the 14 other settings you didn't want included...
Or if you mean that after using that feature, you'd be presented with a checkboxed list of the 90 settings, and you could then check/uncheck those you want included - which is the only practical solution I can think of, but it would be very overwhelming, for the users, but also for us to code to present something usable.

**Comment by zmixnyc** [comment]
Although this Feature Request is closely related to [mine](https://github.com/koreader/koreader/issues/10015), I don't want to hijack this thread too much.

I will try to clarify my request as best I can: 

There are 2 categories of user settings: Global preferences and  Per-Book settings, I am mostly interested in having the ability to save the 21 (Per-Book) user parameters in the lower settings menu and the Font selection from the upper menu to a user accessible file.

Here are the 21 parameters available in the Lower menu

Rotation
Two Columns On/Off

L/R Margins
Sync T/B Margins
Top Margin
Bottom Margin

View Mode
Render Mode
Zoom (dpi)
Line Spacing

Font Size
Word Spacing
Word Expansion

Contrast
Font Weight
Font Hinting
Font Kerning

Alt Status Bar - (I think the  this should be in the upper menu as a Global preference.)
Embedded Style
Embedded Fonts
Image Scaling


The reason I am requesting this feature is that I have found that these parameters need to be adjusted whenever you change to a different Font.  Most of the features in the upper menu are universal enough that they can be globally applied and not need any further adjustments.  "Font" is the exception. Changing the Font requires at least 11 of the lower menu parameters to be adjusted to optimize the layout for the selected font:

L/R Margins
Sync T/B Margins
Top Margin
Bottom Margin
Line Spacing

Font Size
Word Spacing
Word Expansion

Contrast
Font Weight
Font Hinting
Font Kerning

There may be several different solutions to this issue, maybe the 11 items above can be set as Font specific defaults?
That would allow the use to change fonts without concern for the changes required to maintain a consistent layout.

**Comment by solarkite** [comment]
@poire-z , Thank you for your response. Your argument regarding the uniqueness of each book is precisely why I requested original FR. Your own words have made the case for the FR.  Each book that I read has idiosyncrasies that require nuanced setting tweaks. Most people would make those adjustments in the lower menu as designed for the convenience of the user. However, when switching books those settings or not set for the nuances of the new book. With the power of  profiles/gestures one can refine the settings. But it is not linked to the user friendly settings originally laid out. In fact we have a whole different set of settings menus that are not linked. By implementing a mass save function into a profile and/ or gesture we can save the user the complication/ frustration of guessing the settings in a whole other interface.  What ,@poire-z, your main argument is against seems to be the use of quick menus. However, I and others have found it a useful function. And certainly not a bag of crap. The breadth of user functions in the menu systems of koreader has expanded substantially. Just as smartphones have adopted quick menus as menu choices have expanded so to does it make sense for koreader to have functions that facilitate usability. @hius07 has laid out a plan that makes this FR usable, not overwhelming, and appreciative of the nuances you have rightly argued for, as also I have argued for. I’m sorry we seem to disagree about “quick menus” efficacy. I believe the function solves more problems than it creates. @zmixnyc thanks for you voice explaining yours’, and my, and many other’s desire to improve the situation regarding the FR.

**Comment by hius07** [comment]
New screenshots in https://github.com/koreader/koreader/pull/10036.
